src/rocks_compaction_scheduler.cpp

```diff
@@ -61,12 +69,69 @@ namespace mongo {
         if (schedule) {
             log() << "Scheduling compaction to clean up tombstones for prefix "
                   << rocksdb::Slice(prefix).ToString(true);
-            // we schedule compaction now
-            std::string nextPrefix(rocksGetNextPrefix(prefix));
-            rocksdb::Slice begin(prefix), end(nextPrefix);
-            // ignore error
-            rocksdb::experimental::SuggestCompactRange(_db, &begin, &end);
+            // we schedule compaction now (ignoring error)
+            compactPrefix(prefix);
+        }
+    }
+
+    RocksCompactionScheduler::~RocksCompactionScheduler() {
+        {
+            boost::mutex::scoped_lock lk(_compactionMutex);
+            _compactionThreadRunning = false;
+            for (auto item: _compactionQueue) {
+                delete item;
+            }
+        }
+        CancelAllBackgroundWork(_db);
+        _compactionWakeUp.notify_one();
+        _compactionThread.join();
+    }
+
+    void RocksCompactionScheduler::compactionThread() {
+        boost::mutex::scoped_lock lk(_compactionMutex);
+        while (_compactionThreadRunning) {
+            // check if we have something to compact
+            if (_compactionQueue.empty())
+                _compactionWakeUp.wait(lk);
+            else {
+                // get item from queue
+                std::unique_ptr<RocksCompactionOperation> op(_compactionQueue.front());
+                _compactionQueue.pop_front();
+                // unlock mutex for the time of compaction
+                boost::reverse_lock<decltype(lk)> rlk(lk);
+                // do compaction
+                op->run(_db);
+                op.reset();
+            }
         }
+        lk.unlock();
+        LOG(1) << "RocksCompactionScheduler::compactionThread terminating" << std::endl;
     }
 
+    Status RocksCompactionScheduler::scheduleCompactOp(RocksCompactionOperation* cop) {
+        {
+            boost::mutex::scoped_lock lk(_compactionMutex);
+            _compactionQueue.push_back(cop);
+        }
+        _compactionWakeUp.notify_one();
+        return Status::OK();
+    }
+
+    Status RocksCompactionScheduler::compactRange(const rocksdb::Slice* start, const rocksdb::Slice* end) {
+        return scheduleCompactOp(new RocksCompactOpRange(start, end));
+    }
+
+    Status RocksCompactionScheduler::compactPrefix(const std::string& prefix) {
+        rocksdb::Slice start(prefix), end(rocksGetNextPrefix(prefix));
```

Comment:
Temporary object returned by `rocksGetNextPrefix` is destroyed after this line causing dangling ref in `end`.